src/sse/sse_utils.rs

```diff
@@ -0,0 +1,265 @@
+}
+
+// Invert sign of high (2nd) complex
+// values: a.re, a.im, b.re, b.im
+// -->  a.re, a.im, -b.re, -b.im
+#[inline(always)]
+pub unsafe fn negate_hi_f32(values: __m128) -> __m128 {
+    _mm_xor_ps(values, _mm_set_ps(-0.0, -0.0, 0.0, 0.0))
+}
+
+// Duplicate low (1st) complex
+// values: a.re, a.im, b.re, b.im
+// --> a.re, a.im, a.re + a.im
+#[inline(always)]
+pub unsafe fn duplicate_lo_f32(values: __m128) -> __m128 {
+    _mm_shuffle_ps(values, values, 0x44)
+}
+
+// Duplicate high (2nd) complex
+// values: a.re, a.im, b.re, b.im
+// --> b.re, b.im, b.re + b.im
+#[inline(always)]
+pub unsafe fn duplicate_hi_f32(values: __m128) -> __m128 {
+    _mm_shuffle_ps(values, values, 0xEE)
+}
+
+// transpose a 2x2 complex matrix given as [x0, x1], [x2, x3]
+// result is [x0, x2], [x1, x3]
+#[inline(always)]
+pub unsafe fn transpose_complex_2x2_f32(left: __m128, right: __m128) -> [__m128; 2] {
+    let temp02 = extract_lo_lo_f32(left, right);
+    let temp13 = extract_hi_hi_f32(left, right);
+    [temp02, temp13]
+}
+
+// Complex multiplication.
+// Each input contains two complex values, which are multiplied in parallel.
+#[inline(always)]
+pub unsafe fn mul_complex_f32(left: __m128, right: __m128) -> __m128 {
+    //SSE3, taken from Intel performance manual
+    let mut temp1 = _mm_shuffle_ps(right, right, 0xA0);
+    let mut temp2 = _mm_shuffle_ps(right, right, 0xF5);
+    temp1 = _mm_mul_ps(temp1, left);
+    temp2 = _mm_mul_ps(temp2, left);
+    temp2 = _mm_shuffle_ps(temp2, temp2, 0xB1);
+    _mm_addsub_ps(temp1, temp2)
+}
+
+//  __  __       _   _                __   _  _   _     _ _
+// |  \/  | __ _| |_| |__            / /_ | || | | |__ (_) |_
+// | |\/| |/ _` | __| '_ \   _____  | '_ \| || |_| '_ \| | __|
+// | |  | | (_| | |_| | | | |_____| | (_) |__   _| |_) | | |_
+// |_|  |_|\__,_|\__|_| |_|          \___/   |_| |_.__/|_|\__|
+//
+
+pub(crate) struct Rotate90F64 {
+    sign: __m128d,
+}
+
+impl Rotate90F64 {
+    pub fn new(positive: bool) -> Self {
+        let sign = unsafe {
+            if positive {
+                _mm_set_pd(0.0, -0.0)
+            } else {
+                _mm_set_pd(-0.0, 0.0)
+            }
+        };
+        Self { sign }
+    }
+
+    #[inline(always)]
+    pub unsafe fn rotate(&self, values: __m128d) -> __m128d {
+        let temp = _mm_shuffle_pd(values, values, 0x01);
+        _mm_xor_pd(temp, self.sign)
+    }
+}
+
+#[inline(always)]
+pub unsafe fn mul_complex_f64(left: __m128d, right: __m128d) -> __m128d {
+    // SSE3, taken from Intel performance manual
+    let mut temp1 = _mm_unpacklo_pd(right, right);
+    let mut temp2 = _mm_unpackhi_pd(right, right);
+    temp1 = _mm_mul_pd(temp1, left);
+    temp2 = _mm_mul_pd(temp2, left);
+    temp2 = _mm_shuffle_pd(temp2, temp2, 0x01);
+    _mm_addsub_pd(temp1, temp2)
+}
+
+#[cfg(test)]
+mod unit_tests {
+    use super::*;
+    use num_complex::Complex;
+
+    #[inline(always)]
+    unsafe fn mul_complex_f64(left: __m128d, right: __m128d) -> __m128d {
```

Comment:
Does this still need to exist? This feels like test code that might be shadowing the actual mul_64 function.

Reply:
Not needed! Just forgotten there.